BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch

```diff
@@ -0,0 +1,188 @@
+-                return;
++        int varintEnd = in.forEachByte(reader);
++        if (varintEnd == -1) {
++            // We tried to go beyond the end of the buffer. This is probably a good sign that the
++            // buffer was too short to hold a proper varint.
++            if (reader.getResult() == DecodeResult.RUN_OF_ZEROES) {
++                // Special case where the entire packet is just a run of zeroes. We ignore them all.
++                in.clear();
+             }
++            return;
++        }
+ 
+-            // Waterfall start
+-            byte read = in.readByte();
+-            if ( read >= 0 )
+-            {
+-                in.resetReaderIndex();
+-                int length = DefinedPacket.readVarInt( in );
+-                // Waterfall end
+-                if ( false && length == 0) // Waterfall - ignore
+-                {
+-                    throw new CorruptedFrameException( "Empty Packet!" );
+-                }
+-
+-                if ( in.readableBytes() < length )
+-                {
+-                    in.resetReaderIndex();
+-                    return;
+-                    // Waterfall start
+-                } else {
+-                    out.add(in.readRetainedSlice(length));
+-                    return;
+-                    // Waterfall end
++        if (reader.getResult() == DecodeResult.RUN_OF_ZEROES) {
++            // this will return to the point where the next varint starts
++            in.readerIndex(varintEnd);
++        } else if (reader.getResult() == DecodeResult.SUCCESS) {
++            int readVarint = reader.getReadVarint();
++            int bytesRead = reader.getBytesRead();
++            if (readVarint < 0) {
++                in.clear();
++                throw BAD_LENGTH_CACHED;
++            } else if (readVarint == 0) {
++                // skip over the empty packet(s) and ignore it
++                in.readerIndex(varintEnd + 1);
++            } else {
++                int minimumRead = bytesRead + readVarint;
++                if (in.isReadable(minimumRead)) {
++                    out.add(in.retainedSlice(varintEnd + 1, readVarint));
++                    in.skipBytes(minimumRead);
+                 }
+             }
++        } else if (reader.getResult() == DecodeResult.TOO_BIG) {
++            in.clear();
++            throw VARINT_BIG_CACHED;
+         }
+-
+-        throw new CorruptedFrameException( "length wider than 21-bit" );
+     }
+-}
++}
+\ No newline at end of file
+diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/VarintByteDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/VarintByteDecoder.java
+new file mode 100644
+index 00000000..2e4ccf05
+--- /dev/null
++++ b/protocol/src/main/java/net/md_5/bungee/protocol/VarintByteDecoder.java
+@@ -0,0 +1,68 @@
++/*
++ * Courtesy of Velocity
++ * See https://github.com/VelocityPowered/Velocity/commit/5ceac16a821ea35572ff11412ace8929fd06e278
++ */
```

Comment:
I see, this commit is before Velocity's implementation changed to GPL.

Reply:
> I see, this commit is before Velocity's implementation changed to GPL.

Yes, i saw even on other patches of Waterfall same